File: terminal_ticker/agent/loop.py
"""Agent Loop 核心引擎：编排 用户输入 → 模型推理 → 工具调用 → 循环。

参考 OpenAI Codex 和 Nous Hermes 的 agent loop 设计：
- 每轮：构建 prompt → 模型推理 → 判断响应类型
  - 文本响应 → 返回（turn 结束）
  - tool_calls → 执行工具 → 追加结果到历史 → 重新推理（循环）
- 迭代预算防止无限循环
- 对话历史保持 OpenAI message 格式
- 工具结果按调用顺序追加
"""
from __future__ import annotations

import logging
import time
from dataclasses import dataclass, field
from typing import Any, Protocol

from .tools import ToolCall, ToolRegistry, ToolResult

logger = logging.getLogger(__name__)

DEFAULT_MAX_ITERATIONS = 10
DEFAULT_SYSTEM_PROMPT = """你是一个本地运行的 price action trading assistant。
你可以调用工具来获取实时行情、K 线和策略信号。基于工具返回的真实数据做分析。
你不下单、不管理仓位、不承诺收益，也不把分析表述成确定性金融建议。

分析完成后，输出一个 JSON object，字段必须是：
summary: string
bias: "bullish" | "bearish" | "neutral" | "mixed"
confidence: integer 0-100
key_levels: array of {label: string, price: number | null, reason: string}
watch_plan: array of string
invalidation: string
risk_notes: array of string"""


class AgentLLMProvider(Protocol):
    """Agent loop 要求 LLM provider 实现的接口。"""

    name: str
    model: str

    async def chat(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> "ChatResponse": ...


@dataclass
class ChatResponse:
    """LLM provider 的 chat 响应。"""

    content: str | None = None
    tool_calls: list[ToolCall] = field(default_factory=list)
    finish_reason: str = "stop"
    usage: dict[str, int] = field(default_factory=dict)


@dataclass
class LoopStep:
    """Agent loop 中的一步。"""

    step_type: str  # "tool_call" | "tool_result" | "assistant"
    tool_call: ToolCall | None = None
    tool_result: ToolResult | None = None
    content: str | None = None
    timestamp: float = 0.0

    def to_payload(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "stepType": self.step_type,
            "timestamp": self.timestamp,
        }
        if self.tool_call is not None:
            payload["toolCall"] = {
                "id": self.tool_call.id,
                "name": self.tool_call.name,
                "arguments": self.tool_call.arguments,
            }
        if self.tool_result is not None:
            payload["toolResult"] = {
                "callId": self.tool_result.call_id,
                "name": self.tool_result.name,
                "output": self.tool_result.output[:2000],
                "error": self.tool_result.error,
            }
        if self.content is not None:
            payload["content"] = self.content
        return payload


@dataclass
class LoopResult:
    """Agent loop 执行完成的完整结果。"""

    content: str
    steps: list[LoopStep] = field(default_factory=list)
    iterations: int = 0
    total_tokens: int = 0
    finished: bool = True
    error: str | None = None

    def to_payload(self) -> dict[str, Any]:
        return {
            "content": self.content,
            "steps": [step.to_payload() for step in self.steps],
            "iterations": self.iterations,
            "totalTokens": self.total_tokens,
            "finished": self.finished,
            "error": self.error,
        }


class AgentLoop:
    """编排 LLM 推理和工具调用的核心循环。"""

    def __init__(
        self,
        *,
        provider: AgentLLMProvider,
        tools: ToolRegistry,
        system_prompt: str | None = None,
        max_iterations: int = DEFAULT_MAX_ITERATIONS,
    ) -> None:
        self.provider = provider
        self.tools = tools
        self.system_prompt = system_prompt or DEFAULT_SYSTEM_PROMPT
        self.max_iterations = max_iterations

    async def run(
        self,
        user_message: str,
        conversation_history: list[dict[str, Any]] | None = None,
    ) -> LoopResult:
        """执行 agent loop，直到模型产生文本响应或达到迭代上限。"""
        messages = self._build_messages(user_message, conversation_history)
        tool_schemas = self.tools.openai_tool_schemas()

        steps: list[LoopStep] = []
        total_tokens = 0
        iteration = 0

        while iteration < self.max_iterations:
            iteration += 1
            logger.info(
                "agent loop iteration %d/%d, messages=%d",
                iteration, self.max_iterations, len(messages),
            )

            try:
                response = await self.provider.chat(
                    messages=messages,
                    tools=tool_schemas if tool_schemas else None,
                )
            except Exception as exc:
                logger.error("agent loop provider error: %s", exc)
                return LoopResult(
                    content="",
                    steps=steps,
                    iterations=iteration,
                    total_tokens=total_tokens,
                    finished=False,
                    error=str(exc) or exc.__class__.__name__,
                )

            total_tokens += sum(response.usage.values())

            if not response.tool_calls:
                content = response.content or ""
                steps.append(LoopStep(
                    step_type="assistant",
                    content=content,
                    timestamp=time.time(),
                ))
                return LoopResult(
                    content=content,
                    steps=steps,
                    iterations=iteration,
                    total_tokens=total_tokens,
                )

            assistant_msg: dict[str, Any] = {"role": "assistant", "content": response.content or ""}
            assistant_msg["tool_calls"] = [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {
                        "name": tc.name,
                        "arguments": _serialize_arguments(tc.arguments),
                    },
                }
                for tc in response.tool_calls
            ]
            messages.append(assistant_msg)

            for tc in response.tool_calls:
                steps.append(LoopStep(
                    step_type="tool_call",
                    tool_call=tc,
                    timestamp=time.time(),
                ))

                result = await self.tools.execute(tc)

                steps.append(LoopStep(
                    step_type="tool_result",
                    tool_result=result,
                    timestamp=time.time(),
                ))

                messages.append({
                    "role": "tool",
                    "tool_call_id": tc.id,
                    "content": result.output,
                })

        return LoopResult(
            content="Agent reached maximum iteration limit.",
            steps=steps,
            iterations=iteration,
            total_tokens=total_tokens,
            finished=False,
            error=f"Reached max iterations ({self.max_iterations})",
        )

    def _build_messages(
        self,
        user_message: str,
        conversation_history: list[dict[str, Any]] | None,
    ) -> list[dict[str, Any]]:
        """构建发送给模型的消息列表。"""
        messages: list[dict[str, Any]] = [
            {"role": "system", "content": self.system_prompt},
        ]
        if conversation_history:
            for msg in conversation_history:
                role = msg.get("role", "")
                if role in ("user", "assistant"):
                    messages.append({
                        "role": role,
                        "content": str(msg.get("content", "")),
                    })
        messages.append({"role": "user", "content": user_message})
        return messages


import json as _json


def _serialize_arguments(arguments: dict[str, Any]) -> str:
    return _json.dumps(arguments, ensure_ascii=False, separators=(",", ":"))

File: tradex/agent/loop.ts
import { ToolCall, ToolRegistry, ToolResult } from "./tools/registry.js";
import type { AgentModel } from "./models.js";
import { getApiStream } from "./api_registry.js";

export const DEFAULT_MAX_ITERATIONS = 10;
export const DEFAULT_SYSTEM_PROMPT = `你是一名做加密货币永续合约的职业 trader，擅长 price action 与 Smart Money Concepts，习惯用衍生品数据交叉验证判断。默认中文，结论先于论据；涉及行情、K 线、持仓、成交、新闻的事实判断必须先调工具。`;

// ---- Loop-level hooks ----

/**
 * Context passed to all loop hooks after a turn completes.
 */
export interface TurnContext {
  /** Current iteration (1-based) */
  iteration: number;
  /** Accumulated token usage */
  totalTokens: number;
  promptTokens: number;
  /** The full messages array (system + history + tool results so far) */
  messages: Array<Record<string, unknown>>;
  /** The model's response from this turn */
  lastResponse: ChatResponse;
  /** All steps executed so far */
  steps: LoopStep[];
}

/**
 * Partial update returned from prepareNextTurn.
 * Only provided fields take effect; omitted fields keep their current value.
 */
export interface NextTurnUpdate {
  /** Replace the model for subsequent turns (enables mid-conversation model switching) */
  model?: AgentModel;
  /** Replace the system prompt */
  systemPrompt?: string;
  /** Replace the tool registry */
  tools?: ToolRegistry;
  /** Replace the messages array (e.g. for context compression) */
  messages?: Array<Record<string, unknown>>;
}

/**
 * Loop-level lifecycle hooks.
 * All hooks are optional. When not provided, the loop behaves exactly as before.
 */
export interface LoopHooks {
  /**
   * Called after prepareNextTurn. Return true to gracefully stop the loop.
   * Use for: token budget limits, external stop signals, business-logic gates.
   */
  shouldStop?: (ctx: TurnContext) => boolean | Promise<boolean>;

  /**
   * Called after tool execution completes, before shouldStop.
   * Return a partial update to adjust the next turn's configuration.
   * Use for: model escalation, context compression, dynamic tool exposure.
   */
  prepareNextTurn?: (ctx: TurnContext) => NextTurnUpdate | void | Promise<NextTurnUpdate | void>;

  /**
   * Called after shouldStop (if not stopping). Returned messages are appended
   * to the conversation before the next LLM call.
   * Use for: user mid-conversation steering via WebSocket, injecting system notes.
   */
  getSteeringMessages?: (ctx: TurnContext) => Array<Record<string, unknown>> | Promise<Array<Record<string, unknown>>>;
}

export type StreamDeltaHandler = (delta: string) => Promise<void> | void;
export type AgentEventHandler = (event: Record<string, unknown>) => Promise<void> | void;

/**
 * Legacy interface kept for backward compatibility with memory pipeline.
 * New code should use AgentModel + api_registry directly.
 */
export interface AgentLLMProvider {
  name: string;
  model: string;
  chat(input: { messages: Array<Record<string, unknown>>; tools?: Array<Record<string, unknown>> | null; onDelta?: StreamDeltaHandler | null }): Promise<ChatResponse>;
}

export interface ChatResponse {
  content: string | null;
  toolCalls: ToolCall[];
  finishReason: string;
  usage: Record<string, number>;
}

export interface LoopStep {
  stepType: "tool_call" | "tool_result";
  toolCall?: ToolCall | null;
  toolResult?: ToolResult | null;
  timestamp: number;
}

export interface TranscriptMessage {
  role: string;
  content: string;
  metadata?: Record<string, unknown> | null;
  error?: string | null;
}

export interface LoopResult {
  content: string;
  steps: LoopStep[];
  messages: TranscriptMessage[];
  iterations: number;
  totalTokens: number;
  promptTokens: number;
  finished: boolean;
  error: string | null;
}

/**
 * AgentLoop — drives multi-turn tool-calling conversations.
 *
 * Now takes an AgentModel value object instead of a provider instance.
 * Each LLM call dispatches through the global API registry based on model.api.
 *
 * Alternatively, a legacy AgentLLMProvider can still be passed for backward compat.
 */
export class AgentLoop {
  private currentModel: AgentModel | null;
  readonly legacyProvider: AgentLLMProvider | null;
  private currentTools: ToolRegistry;
  private currentSystemPrompt: string;
  readonly maxIterations: number;
  readonly hooks: LoopHooks;

  constructor(input: { model?: AgentModel | null; provider?: AgentLLMProvider | null; tools: ToolRegistry; systemPrompt?: string | null; maxIterations?: number; hooks?: LoopHooks }) {
    this.currentModel = input.model ?? null;
    this.legacyProvider = input.provider ?? null;
    if (!this.currentModel && !this.legacyProvider) {
      throw new Error("AgentLoop requires either a model or a provider");
    }
    this.currentTools = input.tools;
    this.currentSystemPrompt = input.systemPrompt || DEFAULT_SYSTEM_PROMPT;
    this.maxIterations = input.maxIterations ?? DEFAULT_MAX_ITERATIONS;
    this.hooks = input.hooks ?? {};
  }

  get model(): AgentModel | null { return this.currentModel; }
  get tools(): ToolRegistry { return this.currentTools; }
  get systemPrompt(): string { return this.currentSystemPrompt; }

  private async chat(input: { messages: Array<Record<string, unknown>>; tools?: Array<Record<string, unknown>> | null; onDelta?: StreamDeltaHandler | null }): Promise<ChatResponse> {
    if (this.currentModel) {
      const streamFn = getApiStream(this.currentModel.api);
      return streamFn(this.currentModel, input);
    }
    return this.legacyProvider!.chat(input);
  }

  async run(input: { userMessage: string; conversationHistory?: Array<Record<string, unknown>> | null; eventHandler?: AgentEventHandler | null }): Promise<LoopResult> {
    const messages = this.buildMessages(input.userMessage, input.conversationHistory ?? []);
    const steps: LoopStep[] = [];
    const transcript: TranscriptMessage[] = [];
    let totalTokens = 0;
    let promptTokens = 0;
    await emit(input.eventHandler, { type: "agent_start" });
    for (let iteration = 1; iteration <= this.maxIterations; iteration += 1) {
      await emit(input.eventHandler, { type: "turn_start", iteration });
      try {
        // Resolve tool schemas each iteration (hooks may swap the registry)
        const toolSchemas = this.currentTools.openaiToolSchemas();
        const response = await this.chat({
          messages,
          tools: toolSchemas.length ? toolSchemas : null,
          onDelta: async (delta) => emit(input.eventHandler, { type: "message_update", delta }),
        });
        totalTokens += response.usage.total_tokens ?? response.usage.totalTokens ?? 0;
        promptTokens += response.usage.prompt_tokens ?? response.usage.promptTokens ?? 0;
        if (response.content && response.toolCalls.length === 0) {
          messages.push({ role: "assistant", content: response.content });
          transcript.push({ role: "assistant", content: response.content, metadata: usageMetadata(response.usage) });
        }
        if (response.toolCalls.length === 0) {
          await emit(input.eventHandler, agentEndEvent(response.content || "", totalTokens, promptTokens, null));
          return { content: response.content || "", steps, messages: transcript, iterations: iteration, totalTokens, promptTokens, finished: true, error: null };
        }
        messages.push({ role: "assistant", content: response.content || "", tool_calls: openaiToolCallPayloads(response.toolCalls) });
        for (const call of response.toolCalls) {
          steps.push({ stepType: "tool_call", toolCall: call, timestamp: Date.now() / 1000 });
          await emit(input.eventHandler, { type: "tool_call", toolCall: toolCallPayload(call) });
          const result = await this.currentTools.execute(call);
          steps.push({ stepType: "tool_result", toolResult: result, timestamp: Date.now() / 1000 });
          await emit(input.eventHandler, { type: "tool_result", toolResult: toolResultPayload(result) });
          messages.push({ role: "tool", tool_call_id: result.callId, name: result.name, content: result.output });
        }

        // ---- Loop hooks (fire after tool execution, before next iteration) ----
        const turnCtx: TurnContext = { iteration, totalTokens, promptTokens, messages, lastResponse: response, steps };

        // ② prepareNextTurn — adjust configuration for the next turn
        if (this.hooks.prepareNextTurn) {
          const update = await this.hooks.prepareNextTurn(turnCtx);
          if (update) {
            if (update.model) this.currentModel = update.model;
            if (update.systemPrompt) {
              this.currentSystemPrompt = update.systemPrompt;
              messages[0] = { role: "system", content: update.systemPrompt };
            }
            if (update.tools) this.currentTools = update.tools;
            if (update.messages) messages.splice(0, messages.length, ...update.messages);
          }
        }

        // ① shouldStop — graceful early termination
        if (this.hooks.shouldStop) {
          if (await this.hooks.shouldStop(turnCtx)) {
            const content = response.content || "Stopped by loop hook";
            await emit(input.eventHandler, agentEndEvent(content, totalTokens, promptTokens, null));
            return { content, steps, messages: transcript, iterations: iteration, totalTokens, promptTokens, finished: true, error: null };
          }
        }

        // ③ getSteeringMessages — inject external messages before next turn
        if (this.hooks.getSteeringMessages) {
          const steering = await this.hooks.getSteeringMessages(turnCtx);
          for (const msg of steering) messages.push(msg);
        }

      } catch (error) {
        const text = error instanceof Error ? error.message : String(error);
        await emit(input.eventHandler, { type: "error", error: text });
        await emit(input.eventHandler, agentEndEvent(text, totalTokens, promptTokens, text));
        return { content: text, steps, messages: transcript, iterations: iteration, totalTokens, promptTokens, finished: false, error: text };
      }
    }
    const error = "Agent loop reached max iterations";
    await emit(input.eventHandler, agentEndEvent(error, totalTokens, promptTokens, error));
    return { content: error, steps, messages: transcript, iterations: this.maxIterations, totalTokens, promptTokens, finished: false, error };
  }

  private buildMessages(userMessage: string, history: Array<Record<string, unknown>>): Array<Record<string, unknown>> {
    return [{ role: "system", content: this.currentSystemPrompt }, ...history, { role: "user", content: userMessage }];
  }
}

function toolCallPayload(call: ToolCall): Record<string, unknown> {
  return { id: call.id, name: call.name, arguments: call.arguments };
}

function toolResultPayload(result: ToolResult): Record<string, unknown> {
  return { callId: result.callId, name: result.name, output: result.output.slice(0, 2000), error: result.error };
}

function openaiToolCallPayloads(toolCalls: ToolCall[]): Array<Record<string, unknown>> {
  return toolCalls.map((call) => ({ id: call.id, type: "function", function: { name: call.name, arguments: JSON.stringify(call.arguments) } }));
}

function usageMetadata(usage: Record<string, number>): Record<string, unknown> {
  return { usage };
}

function agentEndEvent(content: string, totalTokens: number, promptTokens: number, error: string | null): Record<string, unknown> {
  return { type: "agent_end", content, totalTokens, promptTokens, error };
}

async function emit(handler: AgentEventHandler | null | undefined, event: Record<string, unknown>): Promise<void> {
  if (!handler) return;
  await handler(event);
}
